### PR TITLE
Experimental/efficient trip generation

### PR DESCRIPTION
**Pull request by hong2223**
- a more efficient implementation of the trip generation function using `groupby().apply()`.
- add 'tqdm' as a required library for progress printing.

### CODE REVIEW

File: trackintel/preprocessing/triplegs.py
Comment (henrymartin1) [inline trackintel/preprocessing/triplegs.py:178]: left_of should be tpls.index.name to allow for flexible index names. 
If we are using the index instead of column names we could switch to pd.join() instead of pd.merge()

File: trackintel/preprocessing/triplegs.py
Comment (henrymartin1) [inline trackintel/preprocessing/triplegs.py:181]: I think the index of `tpls` is left untouched by the transformations above. If you can confirm, than we should delete these lines in order to not to change the index name of the user (I think at the moment we do not require that the index has the name 'id' but I am not sure...).

File: trackintel/preprocessing/triplegs.py
Comment (henrymartin1) [inline trackintel/preprocessing/triplegs.py:228]: I think the `df['user_id'].unique()[0]` is a bit confusing. I am ware that we enter this function after grouping by user id, so couldn't we simply tyke `df['user_id'].iloc[0]` ?
On the other way, we could use the `df['user_id'].unique()[0]` for an assert that could help the readability. 
```python
user_id = df['user_id'].unique()
assert len(user_id) == 1
user_id = user_id[0]
unknown_activity = {'user_id': user_id, 'activity': True, 'id': np.nan}
```

**Review by henrymartin1 — CHANGES_REQUESTED** [review]
Hey @hong2223, 
great work! Really nice extension of the trip generation. I had some questions/comments regarding the update of the id's after the trip generation that you should have a look at before we can merge it.

File: requirements.txt
Comment (hong2223) [reply requirements.txt:12]: the dependency is deleted

File: trackintel/preprocessing/triplegs.py
Comment (hong2223) [reply trackintel/preprocessing/triplegs.py:228]: Replaced and solved

File: requirements.txt
Comment (henrymartin1) [reply requirements.txt]: I am pretty sure that I also already deleted it. I was wondering if it could be that it sneaked back in because of a pull request that was not up to date with the origin.

Comment (hong2223) [reply requirements.txt]: Yes, I also remember you deleted it. 
I was careful in sync every time I create a new branch, but maybe at some point I overlooked

File: trackintel/preprocessing/triplegs.py
Comment (hong2223) [reply trackintel/preprocessing/triplegs.py]: solved, now index names can be arbitrary.

File: trackintel/preprocessing/triplegs.py
Comment (hong2223) [reply trackintel/preprocessing/triplegs.py]: solved. now using 'join' the indexes of tpls and spts are left untouched

**Comment by hong2223** [comment]
the function now accepts flexible index names of 'stps' and 'tpls'; the generated 'trips' has 'id' as the index name.

**Review by henrymartin1 — APPROVED** [review]
Nice!

**Pull request merged**
